drivers/ethernet/dsa_ksz8794.c

```diff
@@ -0,0 +1,579 @@
+ */
+int dsa_port_init(struct device *dev)
+{
+	struct dsa_context *context = dev->driver_data;
+	struct dsa_ksz8794_spi *swspi = &phy_spi;
+
+	if (swspi->is_init) {
+		return 0;
+	}
+
+	dsa_hw_init(NULL);
+	k_delayed_work_init(&context->dsa_work, dsa_delayed_work);
+	k_delayed_work_submit(&context->dsa_work, DSA_STATUS_PERIOD_MS);
+
+	return 0;
+}
+
+/* Generic implementation of writing value to DSA register */
+int dsa_write_reg(uint16_t reg_addr, uint8_t value)
+{
+	struct dsa_ksz8794_spi *swspi = &phy_spi;
+
+	if (!swspi->is_init) {
+		return -ENODEV;
+	}
+
+	dsa_ksz8794_write_reg(swspi, reg_addr, value);
+	return 0;
+}
+
+/* Generic implementation of reading value from DSA register */
+int dsa_read_reg(uint16_t reg_addr, uint8_t *value)
+{
+	struct dsa_ksz8794_spi *swspi = &phy_spi;
+
+	if (!swspi->is_init) {
+		return -ENODEV;
+	}
+
+	dsa_ksz8794_read_reg(swspi, reg_addr, value);
+	return 0;
+}
+
+int dsa_set_mac_table_entry(const uint8_t *mac, uint8_t fw_port,
+			    uint16_t tbl_entry_idx, uint16_t flags)
+{
+	struct dsa_ksz8794_spi *swspi = &phy_spi;
+
+	if (!swspi->is_init) {
+		return -ENODEV;
+	}
+
+	if (flags != 0) {
+		return -EINVAL;
+	}
+
+	dsa_ksz8794_set_static_mac_table(swspi, mac, fw_port, tbl_entry_idx);
+
+	return 0;
+}
+
+int dsa_get_mac_table_entry(uint8_t *buf, uint16_t tbl_entry_idx)
+{
+	struct dsa_ksz8794_spi *swspi = &phy_spi;
+
+	if (!swspi->is_init) {
+		return -ENODEV;
+	}
+
+	dsa_ksz8794_get_static_mac_table(swspi, buf, tbl_entry_idx);
+
+	return 0;
+}
+
+#if defined(CONFIG_DSA_KSZ8794_TAIL_TAGGING)
+#define DSA_KSZ8795_TAIL_TAG_OVERRIDE	BIT(6)
+#define DSA_KSZ8795_TAIL_TAG_LOOKUP	BIT(7)
+
+#define DSA_KSZ8794_EGRESS_TAG_LEN 1
+#define DSA_KSZ8794_INGRESS_TAG_LEN 1
+
+#define DSA_MIN_L2_FRAME_SIZE 64
+#define DSA_L2_FCS_SIZE 4
+
+struct net_pkt *dsa_xmit_pkt(struct net_if *iface, struct net_pkt *pkt)
+{
+	struct ethernet_context *ctx = net_if_l2_data(iface);
+	struct net_eth_hdr *hdr = NET_ETH_HDR(pkt);
+	struct net_linkaddr lladst;
+	struct net_buf *buf;
+	uint8_t port_idx, i;
+	size_t len, pad;
+
+	lladst.len = sizeof(hdr->dst.addr);
+	lladst.addr = &hdr->dst.addr[0];
+
+	len = net_pkt_get_len(pkt);
+	/*
+	 * For KSZ8794 one needs to 'pad' the L2 frame to its minimal size
+	 * (64B) before appending TAIL TAG and FCS
+	 */
+	if (len < (DSA_MIN_L2_FRAME_SIZE - DSA_L2_FCS_SIZE)) {
+		/* Calculate number of bytes needed for padding */
+		pad = DSA_MIN_L2_FRAME_SIZE - DSA_L2_FCS_SIZE - len;
+		buf = net_buf_alloc_len(net_buf_pool_get(pkt->buffer->pool_id),
+					pad, K_NO_WAIT);
+
+		if (!buf) {
+			LOG_ERR("DSA cannot allocate new data buffer");
+			return NULL;
+		}
+
+		buf = net_buf_frag_last(net_buf_frag_add(pkt->buffer, buf));
+		for (i = 0; i < pad; i++)
+			net_buf_add_u8(buf, 0x0);
+	}
+
+	buf = net_buf_alloc_len(net_buf_pool_get(pkt->buffer->pool_id),
+				DSA_KSZ8794_INGRESS_TAG_LEN, K_NO_WAIT);
+	buf = net_buf_frag_last(net_buf_frag_add(pkt->buffer, buf));
```

Comment:
you could avoid allocating 2 net_buf here, just allocate one at the beginning of the function (ask for the max length you need, it's anyway going to fit in actual fixed length of net_buf). Pre-fill it with padding if necessary, then with port_idx.

Reply:
The code as is looks more logical to me as:

1. I first check if the packet is smaller than 64B (if yes - I allocate padding and append it)
2. Then I do allocate always (no matter of the packet size) a single byte for the tag, which logically is not related to preparing buffer for packet, as it is removed in the KSZ8794 switch.
3. The last step is to setup the tag byte to the buffer allocated earlier.

The code as it is works seamlessly when the packet is larger than 64B (minimal packet size), as we would need the extra allocation for tag always. For packet smaller than 64B we would have two buffers allocated and appended (pad and tag) to the original packet.

Reply:
It can be kept logical _and_ memory efficient. Here you'll eat up 256 bytes for at most 65bytes (if net_buf buffer size is 128bytes, it gets worse if that value is higher: 512 is 256 etc etc...).

- calculate the padding
- allocate for the padding and the tag
- if (pad) ... add the padding
- add the tag.

Reply:
Ok, I've rewritten it to use only single `net_buf_alloc_len()` allocation. Thanks for sharing the idea.